heavyball/__init__.py

```diff
@@ -254,6 +266,312 @@ def __init__(self, params, lr=0.001, beta=0.9, weight_decay=0.0, preconditioner_
+                else:
+                    state['diag_hessian'] = h.clone()
+
+                # Set minimum value for stability
+                state['diag_hessian'].clamp_(min=1e-6)
+
+        # 표준 최적화 진행
+        if not group['foreach'] or len(p) == 1:
+            for param, grad in zip(p, g):
+                C.chain(self.state_, group, [grad], [param], *self.fns)
+        else:
+            C.chain(self.state_, group, g, p, *self.fns)
+
+        group['caution'] = caution
+        group['lr'] = group['prev_lr']
+        group['step'] = None
+
+class ForeachSophiaG(ForeachSophiaH):
+    """
+    Sophia optimizer with Gauss-Newton-Bartlett estimator (same implementation as SophiaH for simplicity).
+    """
+    def __init__(self, params, lr=1e-3, betas=(0.96, 0.99), eps=1e-12, weight_decay=0.1,
+                warmup_steps=0, gamma=0.05, update_freq=10, foreach: bool = True,
+                storage_dtype: str = 'float32', mars: bool = False, caution: bool = False,
+                mars_gamma: float = 0.0025, gradient_clipping: C.str_or_fn = C.use_default,
+                update_clipping: C.str_or_fn = C.use_default, palm: bool = C.use_default,
+                beta2_scale: float = 0.8):
+
+        defaults = locals()
+        defaults.pop("self")
+        params_local = defaults.pop("params")
+
+        # Add Sophia-specific parameters to defaults
+        defaults['sophia_gamma'] = defaults.pop('gamma')
+        defaults['sophia_update_freq'] = defaults.pop('update_freq')
+        defaults['sophia_type'] = 'g'  # Gauss-Newton-Bartlett estimator
+
+        C.BaseOpt.__init__(self, params_local, defaults, foreach, gradient_clipping, update_clipping, palm,
+                         C.zero_guard("diag_hessian"),
+                         C.scale_by_sophia)
+
+
+    def estimate_hessian_g(self, state, group, update, grad, param):
+        """
+        Estimate diagonal Hessian using Gauss-Newton-Bartlett method.
+        """
+        k = group.get('sophia_update_freq', 10)
+        beta2 = utils.get_beta2(group)
+        state['hessian_step'] += 1
+
+        # Check if it's time to update the Hessian estimate
+        if state['hessian_step'] >= state['next_hessian_update']:
+            state['next_hessian_update'] = state['hessian_step'] + k
+
+            # For GNB we can use a simplified approximation based on squared gradients
+            # In a real LM, this should be implemented with proper label resampling
+            h = grad * grad  # Simplified GNB estimator
+
+            # Update EMA of diagonal Hessian
+            if 'diag_hessian' in state:
+                state['diag_hessian'].mul_(beta2).add_(h.abs(), alpha=1-beta2)
+            else:
+                state['diag_hessian'] = h.abs().clone()
+
+            # Ensure positive values for numerical stability
+            state['diag_hessian'].clamp_(min=1e-6)
+
+        return update
+
+def _init_adalomo(state, group, update, grad, param):
```

Comment:
could this function be moved to the chainable api / is it necessary here? `__init__` was intended as the main API with only the optimizer definitions